session/auth/authprovider/tokenseed.go

```diff
@@ -0,0 +1,65 @@
+package authprovider
+
+import (
+	"crypto/rand"
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+type tokenSeeds struct {
+	mu  sync.Mutex
+	dir string
+}
+
+type seed struct {
+	Seed []byte
+}
+
+func (ts *tokenSeeds) getSeed(host string) ([]byte, error) {
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+
+	fp := filepath.Join(ts.dir, ".token_seed")
+	dt, err := ioutil.ReadFile(fp)
+	m := map[string]seed{}
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			return nil, err
+		}
+	} else {
+		if err := json.Unmarshal(dt, &m); err != nil {
+			return nil, errors.Wrapf(err, "failed to parse %s", fp)
+		}
+	}
+	v, ok := m[host]
+	if !ok {
+		v = seed{Seed: newSeed()}
+	}
+
+	m[host] = v
+
+	dt, err = json.MarshalIndent(m, "", "  ")
+	if err != nil {
+		return nil, err
+	}
+
+	if err := os.MkdirAll(filepath.Dir(fp), 0701); err != nil {
+		return nil, err
+	}
+
+	if err := ioutil.WriteFile(fp, dt, 0600); err != nil {
```

Comment:
`WriteFile` opens the file w/ `O_TRUNC`, so if the daemon crashed at the exact wrong time, I think this file could end up existing but be empty, which would cause the `json.Unmarshal` above to always return an error. Obviously pretty obscure, but if that happened the only recourse would be for the user to manually delete the file.

Maybe if the json unmarshal fails you can just assume the file is empty and start anew?

Reply:
Would you prefer a file lock?

Reply:
Sure lock would work, or you could write to a new temp file opened w/ `O_EXCL` and `rename(2)` it to the final one once done, whichever is easiest

Reply:
Added file lock